### ISSUE TITLE
[vLLM] Expand and optimize CPU sampling for non-greedy serving

### ISSUE THREAD

**Issue by kmabeeTT** [tenstorrent/tt-xla#4370]
## Summary

Device-side non-greedy sampling is currently ~5 tok/s  (WIP https://github.com/tenstorrent/tt-xla/issues/3940 improvements coming) -- unusable for demos or production. Device sampling performance is being actively worked on separately, but in the near term `cpu_sampling=True` is the practical workaround. We've extended the CPU sampling path to support full non-greedy sampling (temperature, penalties, top-k/top-p) and optimized it with the Gumbel-max trick. These changes have been running in demos for the past 3 weeks.

Note: we also had a fix for the cpu_sampling routing bug (path never being reached) but dropped it since main already has the same fix via https://github.com/tenstorrent/tt-xla/pull/4332.

Planning to put up a PR from branch `kmabee/cpu_sampling_improvements`. The relevant commits:

| Commit | Description |
|---|---|
| `3f215c795` | Extend CPU sampling to support temperature, penalties, and top-k/top-p |
| `7f3848aae` | Optimize CPU sampling: Gumbel-max trick + batched top-k path |

## Performance

Llama-3.1-8B-Instruct, single P150, cpu_sampling=True, bfp_bf8, opt_level=1:

| Batch | Forward-pass ceiling | CPU sampling | Gap |
|---|---|---|---|
| 2 (demo) | 21.7 tok/s | 19.0 tok/s | 12% |
| 8 | 19.7 tok/s | 15.1 tok/s | 23% |

The remaining gap is the device-to-CPU logits transfer. The main optimization was replacing `torch.softmax + torch.multinomial` with the Gumbel-max trick (`argmax(logits + Gumbel noise)`) -- mathematically equivalent but avoids `multinomial` which processes each batch element sequentially.

## Gumbel-max optimization

The main CPU sampling bottleneck was `torch.multinomial` -- it processes each batch element sequentially and is expensive at vocab size 128K. We replaced it with the Gumbel-max trick: adding Gumbel-distributed noise to the logits and taking `argmax`. This is a well-known method for sampling from a categorical distribution without explicitly computing softmax or calling multinomial -- `argmax(logits + Gumbel(0,1))` produces samples from exactly the same distribution as `multinomial(softmax(logits))`.

```python
# Before: softmax + multinomial (~4x slower at vocab=128K)
probs = torch.softmax(logits, dim=-1)
random = torch.multinomial(probs, num_samples=1).squeeze(-1)

# After: Gumbel-max (single batched argmax, same distribution)
gumbel = -torch.log(-torch.log(torch.rand_like(logits.float()) + 1e-20) + 1e-20)
random = torch.argmax(logits + gumbel, dim=-1)
```

This matches what the device-side sampler already does -- `random_sample()` in `sampler.py` uses the equivalent exponential trick (`probs.div_(exponential).argmax()`). Dividing probabilities by Exp(1) random variables and taking argmax is mathematically the same as adding Gumbel noise to log-probabilities and taking argmax. Both avoid `torch.multinomial`.
